service/header/sync.go

```diff
@@ -69,17 +75,57 @@ func (s *Syncer) Start(context.Context) error {
 }
 
 // Stop stops Syncer.
-func (s *Syncer) Stop(context.Context) error {
-	s.cancel()
-	s.cancel = nil
-	return nil
+func (s *Syncer) Stop(ctx context.Context) error {
+	defer func() {
+		s.cancel()
+		s.cancel = nil
+	}()
+	return s.WaitSync(ctx)
 }
 
 // IsSyncing returns the current sync status of the Syncer.
 func (s *Syncer) IsSyncing() bool {
 	return atomic.LoadUint64(&s.inProgress) == 1
 }
 
+// WaitSync blocks until ongoing sync is done.
+func (s *Syncer) WaitSync(ctx context.Context) error {
+	state := s.State()
+	if state.Finished() {
+		return nil
+	}
+
+	// this store method blocks until header is available
+	_, err := s.store.GetByHeight(ctx, state.ToHeight)
+	return err
+}
+
+// SyncState collects all the information about a sync.
+type SyncState struct {
+	ID                   uint64 // incrementing ID of a sync
```

Comment:
ID sounds like it would be globally unique to SyncStates but in fact it's just a counter. Besides that it gets incremented below, what is its purpose?

Reply:
It doesn't really seemed to be used besides getting incremented?

Reply:
Ah NVM, it get's returned as part of the SyncState. Is SyncState/ID sth an RPC endpoint will ever expose or users have to understand in logs? If so. it is even important that the name is clearer: maybe `Counter` instead of `ID` or `Round`.

Reply:
I believe it will be useful for debugging synchronization, so we can differentiate different syncs between each other. I got this experience from debugging my lotus(filecoin) node. So once we will have a cmd that accesses SyncStatus, we will print out to the command line the ID of a sync that might stall, and identify it with ID.

Reply:
This will be accessed from the RPC - yes.
> Is SyncState/ID sth an RPC endpoint will ever expose or users have to understand in logs

Reply:
Why not ID though? It is an identifier of a sync, first of all, so one can identify it. It is also a counter, yes, as we increment it, but the main purpose is identification/diffirention.